### PR TITLE
keep slot connection tags when resolving recipe by verb

### PR DESCRIPTION
**Pull request by mmandlis**
addresses: https://github.com/PolymerLabs/arcs/issues/1133

PS: if this makes, I can do a similar thing for handles.

### CODE REVIEW

File: runtime/strategies/match-recipe-by-verb.js [hunk 1]
```diff
@@ -71,26 +71,27 @@ export default class MatchRecipeByVerb extends Strategy {
 
             particleForReplacing.remove();
 
-
             for (let consumeSlot in slotConstraints) {
-              if (slotConstraints[consumeSlot].targetSlot || Object.values(slotConstraints[consumeSlot].providedSlots).filter(a => a != null).length > 0) {
+              if (slotConstraints[consumeSlot].tags.length > 0 || slotConstraints[consumeSlot].targetSlot || Object.values(slotConstraints[consumeSlot].providedSlots).filter(a => a != null).length > 0) {
```
Comment (shans) [inline runtime/strategies/match-recipe-by-verb.js]: this is supposed to be checking whether there's a concrete slot that's already been mapped by the containing recipe. I don't think you want to add the tags check here?

File: runtime/strategies/match-recipe-by-verb.js [hunk 2]
```diff
@@ -47,7 +47,7 @@ export default class MatchRecipeByVerb extends Strategy {
         let slotConstraints = {};
         for (let consumeSlot of Object.values(particle.consumedSlotConnections)) {
           let targetSlot = consumeSlot.targetSlot && consumeSlot.targetSlot.sourceConnection ? consumeSlot.targetSlot : null;
-          slotConstraints[consumeSlot.name] = {targetSlot, providedSlots: {}};
+          slotConstraints[consumeSlot.name] = {tags: consumeSlot.tags, targetSlot, providedSlots: {}};
           for (let providedSlot of Object.keys(consumeSlot.providedSlots)) {
             let sourceSlot = consumeSlot.providedSlots[providedSlot].consumeConnections.length > 0 ? consumeSlot.providedSlots[providedSlot] : null;
             slotConstraints[consumeSlot.name].providedSlots[providedSlot] = sourceSlot;
```
Comment (shans) [inline runtime/strategies/match-recipe-by-verb.js]: can the provided slots not have tags too?

**Comment by shans** [comment]
If I understand correctly, this acts to accumulate tags into slots that are explicitly matched by name, when substituting recipes for verbs.

i.e.
```
recipe aVerb
  P
    consume foo

recipe
  particle can aVerb
    consume foo #woo

--->

recipe
  P
    consume foo #woo
```

I had something quite different in mind, and I suspect Berni did too. What I'd like to see is something like:

```
recipe aVerb
  P
    consume #annotation #item #subitem
    consume #special_other_thing

recipe
  particle can aVerb
    consume #annotation as s0
  Q
    consume master
      provide as s0

--->

recipe
  P
    consume as s0
    consume #special_other_thing
  Q 
    consume master
      provide as s0
```

in other words, the verb is used to establish the mapping, not accumulated after mapping is already established.

WDYT?

**Comment by mmandlis** [comment]
Ignoring verbs and nested recipes for a second, if i have 2 particles:
```
particle P
  consume slotP

particle Q
  consume slotQ
    provide slotQq

recipe
  P
    consume slotP #details
  Q
     consume slotQ
       provide slotQq #details
```

once resolve-recipe runs and slots are mapped, the resulting recipe will be
```
recipe
  P
    consume slotP #details as slot0
  Q
     consume slotQ
       provide slotQq #details as slot0
```
and the assigned tag will be kept for the slot, because it is useful information to have for the next steps.

is this not a behavior you'd expect? otherwise, why it would be different for nested recipes? 
i think that if a slot (or slot-connection) is given a tag at some point, this tag should stick, because it is describing the slot and important for future recipes.

**Comment by shans** [comment]
What should this do?

```
recipe aVerb
  P
    consume annotation
  Q
    consume annotation #foo

recipe
  particle can aVerb
    consume #foo as s0
  R
    consume master
      provide as s0
````

I agree that, once matched, excess tags should be copied across. However, I think this should be secondary to using tags to match _which_ slot to bind to.

**Pull request closed without merging**